File: csharp/metafiles/metafiles.cs
// ============================================================================
//  metafiles -- C# port of examples\Vb6\metafiles\metafiles.bas
//  Places three EMF metafiles, each centered and scaled to a landscape page,
//  with a red frame around them.
// ============================================================================
using System;
using System.IO;
using LumasPdfSdk;

class Metafiles
{
    const uint CLR_RED = 255;
    const double MARGIN = 10.0;

    static TErrorProc _err = ErrProc;

    static int ErrProc(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine(errMessage);
        return 0;
    }

    static void PlaceEMFCentered(IntPtr pdf, string mFile, double width, double height)
    {
        TRectL r = new TRectL();
        LumasPdf.pdfGetLogMetafileSizeW(pdf, mFile, ref r);
        double w = r.Right - r.Left;
        double h = r.Bottom - r.Top;
        width = width - 2.0 * MARGIN;
        height = height - 2.0 * MARGIN;
        double sx = width / w;

        double x, y;
        if (h * sx <= height)
        {
            x = MARGIN;
            h = h * sx;
            y = (height - h) / 2.0;
            LumasPdf.pdfInsertMetafileW(pdf, mFile, x, y, width, 0.0);
            LumasPdf.pdfSetStrokeColor(pdf, CLR_RED);
            LumasPdf.pdfRectangle(pdf, x, y, width, h, (int)TPathFillMode.fmStroke);
        }
        else
        {
            sx = height / h;
            w = w * sx;
            x = (width - w) / 2.0;
            y = MARGIN;
            LumasPdf.pdfInsertMetafileW(pdf, mFile, x, y, 0.0, height);
            LumasPdf.pdfSetStrokeColor(pdf, CLR_RED);
            LumasPdf.pdfRectangle(pdf, x, y, w, height, (int)TPathFillMode.fmStroke);
        }
    }

    static void Main()
    {
        string exeDir = AppDomain.CurrentDomain.BaseDirectory;

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        if (!LumasPdf.pdfCreateNewPDFW(pdf, ""))       // output file opened later
        {
            LumasPdf.pdfDeletePDF(pdf);
            return;
        }

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        string[] emfs = { "coords.emf", "fulltest.emf", "gdi.emf" };
        foreach (string emf in emfs)
        {
            LumasPdf.pdfAppend(pdf);
            LumasPdf.pdfSetOrientationEx(pdf, 90);
            PlaceEMFCentered(pdf, Path.Combine(exeDir, emf),
                LumasPdf.pdfGetPageWidth(pdf), LumasPdf.pdfGetPageHeight(pdf));
            LumasPdf.pdfEndPage(pdf);
        }

        string outFile = Path.Combine(exeDir, "out.pdf");
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
        }
        if (LumasPdf.pdfCloseFile(pdf))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: c/metafiles/metafiles.c
/* metafiles -- C port of examples\Vb6\metafiles\metafiles.bas
 * Places three EMF metafiles centered/scaled to landscape pages with a red frame. */
#include <stdio.h>
#include "lumaspdf.h"
#include "../_common.h"

#define CLR_RED 255
#define MARGIN 10.0

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static void PlaceEMFCentered(PPDF pdf, const char* MFile, double Width, double Height)
{
    double x, y, w, h, sx;
    TRectL r;

    pdfGetLogMetafileSizeA(pdf, MFile, &r);
    w = (double)(r.Right - r.Left);
    h = (double)(r.Bottom - r.Top);
    Width -= 2.0 * MARGIN;
    Height -= 2.0 * MARGIN;
    sx = Width / w;

    if (h * sx <= Height) {
        x = MARGIN;
        h = h * sx;
        y = (Height - h) / 2.0;
        pdfInsertMetafileA(pdf, MFile, x, y, Width, 0.0);
        pdfSetStrokeColor(pdf, CLR_RED);
        pdfRectangle(pdf, x, y, Width, h, fmStroke);
    } else {
        sx = Height / h;
        w = w * sx;
        x = (Width - w) / 2.0;
        y = MARGIN;
        pdfInsertMetafileA(pdf, MFile, x, y, 0.0, Height);
        pdfSetStrokeColor(pdf, CLR_RED);
        pdfRectangle(pdf, x, y, w, Height, fmStroke);
    }
}

int main(int argc, char** argv)
{
    PPDF pdf;
    char dir[1024], outFile[1100], f1[1100], f2[1100], f3[1100];
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    if (pdfCreateNewPDFA(pdf, "") == 0) { pdfDeletePDF(pdf); return 1; }
    pdfSetPageCoords(pdf, pcTopDown);
    exedir(argv[0], dir, sizeof(dir));
    _snprintf(f1, sizeof(f1), "%s\\coords.emf", dir);
    _snprintf(f2, sizeof(f2), "%s\\fulltest.emf", dir);
    _snprintf(f3, sizeof(f3), "%s\\gdi.emf", dir);
    _snprintf(outFile, sizeof(outFile), "%s\\out.pdf", dir);

    pdfAppend(pdf); pdfSetOrientationEx(pdf, 90);
    PlaceEMFCentered(pdf, f1, pdfGetPageWidth(pdf), pdfGetPageHeight(pdf));
    pdfEndPage(pdf);

    pdfAppend(pdf); pdfSetOrientationEx(pdf, 90);
    PlaceEMFCentered(pdf, f2, pdfGetPageWidth(pdf), pdfGetPageHeight(pdf));
    pdfEndPage(pdf);

    pdfAppend(pdf); pdfSetOrientationEx(pdf, 90);
    PlaceEMFCentered(pdf, f3, pdfGetPageWidth(pdf), pdfGetPageHeight(pdf));
    pdfEndPage(pdf);

    if (pdfHaveOpenDoc(pdf)) {
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 1; }
    }
    if (pdfCloseFile(pdf))
        printf("PDF file \"%s\" successfully created!\n", outFile);
    pdfDeletePDF(pdf);
    return 0;
}
